transactionoutbox-cdi/src/main/java/com/gruelbox/transactionoutbox/CdiTransactionManager.java

```diff
@@ -0,0 +1,174 @@
+package com.gruelbox.transactionoutbox;
+
+import static com.gruelbox.transactionoutbox.Utils.uncheck;
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+
+import javax.sql.DataSource;
+import javax.transaction.Status;
+import javax.transaction.Synchronization;
+import javax.transaction.TransactionSynchronizationRegistry;
+import javax.transaction.Transactional;
+import javax.transaction.Transactional.TxType;
+
+
+public class CdiTransactionManager implements ThreadLocalContextTransactionManager
+{
+   private final CdiTransaction transactionInstance;
+
+   private TransactionSynchronizationRegistry tsr;
+
+   public CdiTransactionManager(DataSource datasource, TransactionSynchronizationRegistry tsr) throws SQLException
+   {
+      transactionInstance = new CdiTransaction(datasource);
+      this.tsr = tsr;
+   }
+
+   @Override
+   @Transactional(value = TxType.REQUIRES_NEW)
+   public void inTransaction(Runnable runnable)
+   {
+      uncheck(() -> inTransactionReturnsThrows(ThrowingTransactionalSupplier.fromRunnable(runnable)));
+   }
+
+   @Override
+   @Transactional(value = TxType.REQUIRES_NEW)
+   public void inTransaction(TransactionalWork work)
+   {
+      uncheck(() -> inTransactionReturnsThrows(ThrowingTransactionalSupplier.fromWork(work)));
+   }
+
+   @Override
+   @Transactional(value = TxType.REQUIRES_NEW)
+   public <T, E extends Exception> T inTransactionReturnsThrows(ThrowingTransactionalSupplier<T, E> work) throws E
+   {
+      return work.doWork(transactionInstance);
+   }
+
+   @Override
+   public <T, E extends Exception> T requireTransactionReturns(ThrowingTransactionalSupplier<T, E> work) throws E, NoTransactionActiveException
+   {
+      if (tsr.getTransactionStatus() != Status.STATUS_ACTIVE)
+      {
+         throw new NoTransactionActiveException();
+      }
+
+      return work.doWork(transactionInstance);
+   }
+
+   private final class CdiTransaction implements Transaction
+   {
+
+      private DataSource datasource;
+
+      public CdiTransaction(DataSource datasource) throws SQLException
+      {
+         this.datasource = datasource;
+      }
+
+      @Override
+      public final Connection connection()
+      {
+         try
+         {
+            return datasource.getConnection();
```

Comment:
I'm a little unclear on this: By default, I'd expect a `DataSource` to return a new connection every time (or at least an available connection from a pool) whereas here it looks like you're assuming it only returns the active connection in the current transaction.

In Quarkus, is the injectable `DataSource` a proxy that ties into the current active transaction?

Reply:
If I remember correctly from previous roles working with `TransactionSynchronizationRegistry`, you can get the currently active connection from there...

Reply:
Can't find another way to get the active connection using only API methods

Reply:
As long as it is definitely the active connection you get here, all good